ruby/ql/lib/codeql/ruby/security/KernelOpenQuery.qll

```diff
@@ -16,21 +16,70 @@ class AmbiguousPathCall extends DataFlow::CallNode {
     this.(KernelMethodCall).getMethodName() = "open" and
     name = "Kernel.open"
     or
-    this = API::getTopLevelMember("IO").getAMethodCall("read") and
-    not this = API::getTopLevelMember("File").getAMethodCall("read") and // needed in e.g. opal/opal, where some calls have both paths, but I'm not sure why
+    methodCallOnlyOnIO(this, "read") and
     name = "IO.read"
+    or
+    methodCallOnlyOnIO(this, "write") and
+    name = "IO.write"
+    or
+    methodCallOnlyOnIO(this, "binread") and
+    name = "IO.binread"
+    or
+    methodCallOnlyOnIO(this, "binwrite") and
+    name = "IO.binwrite"
+    or
+    methodCallOnlyOnIO(this, "foreach") and
+    name = "IO.foreach"
+    or
+    methodCallOnlyOnIO(this, "readlines") and
+    name = "IO.readlines"
+    or
+    this = API::getTopLevelMember("URI").getAMethodCall("open") and
+    name = "URI.open"
   }
 
   /** Gets the name for the method being called. */
   string getName() { result = name }
 
   /** Gets the name for a safer method that can be used instead. */
   string getReplacement() {
+    result = "File.open" and name = "Kernel.open"
+    or
     result = "File.read" and name = "IO.read"
     or
-    result = "File.open" and name = "Kernel.open"
+    result = "File.write" and name = "IO.write"
+    or
+    result = "File.binread" and name = "IO.binread"
+    or
+    result = "File.binwrite" and name = "IO.binwrite"
+    or
+    result = "File.foreach" and name = "IO.foreach"
+    or
+    result = "File.readlines" and name = "IO.readlines"
+    or
+    result = "URI(<uri>).open" and name = "URI.open"
   }
 
   /** Gets the argument that specifies the path to be accessed. */
   DataFlow::Node getPathArgument() { result = this.getArgument(0) }
 }
+
+private predicate methodCallOnlyOnIO(DataFlow::CallNode node, string methodName) {
+  node = API::getTopLevelMember("IO").getAMethodCall(methodName) and
+  not node = API::getTopLevelMember("File").getAMethodCall(methodName) // needed in e.g. opal/opal, where some calls have both paths (opal implements an own corelib)
+}
+
+/**
+ * A sanitizer for kernel open vulnerabilities.
+ */
+abstract class Sanitizer extends DataFlow::Node { }
+
+/**
+ * If a File.join is performed the resulting string will not start with a pipe |
+ * This is true as long the tainted data doesn't flow into the first argument.
+ */
+private class FileJoinCall extends Sanitizer, DataFlow::ExprNode {
```

Comment:
Marking the arguments here ends up potentially sanitizing nodes on a dataflow path that's not actually sanitized, e.g. in 
```
 file = params[:file]
 IO.read(File.join("", file)) # this call is safe
 IO.read(file)                # this call is unsafe, but marking the `file` argument to `File.join` as a sanitizer means that we think it's safe
```
With the `SummarizedCallable::getParameter` predicate from https://github.com/github/codeql/pull/11621, we can redefine this sanitizer as:
```
private class FileJoinSanitizer extends Sanitizer {
  FileJoinSanitizer() { this = any(File::FileJoinSummary s).getParameter("1..") }
}
```
after changing https://github.com/github/codeql/blob/6fa2fe6c864d522769fa3132b0b9c869775a53b8/ruby/ql/lib/codeql/ruby/frameworks/Files.qll#L136 to `input = "Argument[0,1..]"`.

This lets us use the synthesized parameter node (for any parameter with index 1 or higher) as a sanitizer, meaning that only flow into the `File.join` call is sanitized rather than all flow from the argument.

I recommend trying this out with a new test case to check that I haven't made a mistake here.

Reply:
The new tests look good 👍